<a href="https://colab.research.google.com/github/TsamayaDesigns/codeDivision-data-with-python/blob/main/Describing_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Describing data with summary statistics
---

This worksheet is tied to the Pandas Getting Started Tutorials, picking out particular tutorials to link them into a theme here.

We will focus on describing data.  This is the least risky in terms of bias and inaccurate conclusions as it should focus just on what data is presented to us.

Each exercise will ask you to work through on tutorial on the Getting Started page, to try the code from the tutorial here and to try a second, similar action.

---

The practice data from the tutorials comes from a dataset on Titanic passengers.


### Exercise 1 - open the Titanic dataset
---

The Titanic dataset is stored at this URL:
https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Read the dataset into a pandas dataframe that you will call **titanic**.

**Test output**:  
The shape of the dataframe will be (891, 12)

In [31]:
import pandas as pd

# Import CSV-data from GitHub
def get_csv_data():
  url = " https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Exercise 1
print(data.shape)

(891, 12)


### Exercise 2 - get summary information about the dataframe
---

Read through the tutorials:  
[What kind of data does pandas handle?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#)  
[How do I read and write tabular data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html)

Use panda functions to display the following:
1.  A technical summary of the data (info())
2.  A description of the numerical data (describe())
3.  Display the Series 'Age'

**Test output**:   
1.  The info should show that there are only 204 values in the Cabin series, out of 891 records.  
2.  The description should show 7 columns and a mean age of 29.699118
3.  The Age series should have values of type float64 and Length 891

In [32]:
# Exercise 2
# 1. info()
tech_summary = data["Cabin"].info()
print("\nTechnical Summary: \n", tech_summary)

# 2. describe()
data_descript = data.describe()
print("\nData Description: \n", data_descript)

# 3. Display the Series 'Age'
data_age = data["Age"]
print("\nDisplay the Series 'Age': \n", data_age)

<class 'pandas.core.series.Series'>
RangeIndex: 891 entries, 0 to 890
Series name: Cabin
Non-Null Count  Dtype 
--------------  ----- 
204 non-null    object
dtypes: object(1)
memory usage: 7.1+ KB

Technical Summary: 
 None

Data Description: 
        PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.6934

### Exercise 3 - aggregating statistics
---

Read through the tutorial:  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

Use panda functions to display the following summary statistics from the titanic dataset:  

1.  The average (mean) age of passengers  
2.  The median age and fare  
3.  The mean fare
4.  The modal fare and gender

**Test output**:   
29.699118, Age 28.0000 Fare 14.4542, 32.2042079685746, Fare 8.05 Sex male


In [33]:
# Exercise 3
# 1. The average (mean) age of passengers
pass_mean_age = data["Age"].mean()
print("\nDisplay (mean) age of passengers: \n", pass_mean_age)

# 2. The median age and fare
median_age_fare = data[["Age", "Fare"]].median()
print("\nDisplay median age and fare: \n", median_age_fare)

# 3. The mean fare
mean_fare = data["Fare"].mean()
print("\nDisplay (mean) age of passengers: \n", mean_fare)

# 4. The modal "mode()" fare and gender
modal_fare_gender = data[["Fare", "Sex"]].mode()
print("\nDisplay modal fare and gender: \n", modal_fare_gender)



Display (mean) age of passengers: 
 29.69911764705882

Display median age and fare: 
 Age     28.0000
Fare    14.4542
dtype: float64

Display (mean) age of passengers: 
 32.204207968574636

Display modal fare and gender: 
    Fare   Sex
0  8.05  male


### Exercise 4 - displaying other statistics
---

Take a look at the list of methods available for giving summary statistics [here](https://pandas.pydata.org/docs/user_guide/basics.html#basics-stats)

Use panda functions, and your existing knowledge, to display the following summary statistics from the titanic dataset:

1.  The total number of passengers on the titanic
2.  The age of the youngest passenger
3.  The most expensive ticket price
4.  The range of ticket prices
5.  The number of passenges with cabins
6.  The code for the port where the highest number of passengers embarked
7.  The most populous gender
8.  The standard deviation for age and fare

**Tests**:  
891, 0.42, 512.3292, 512.3292, 204, S, male, Age 14.526497 Fare 49.693429

In [34]:
def investigate_data(data):
  # The total number of passengers on the titanic
  tot_pass = data["PassengerId"].count()

  # The age of the youngest passenger
  youngest_pass = data["Age"].min()

  # The most expensive ticket price
  most_exp_ticket = data["Fare"].max()

  # The range of ticket prices
  ticket_price_ranges = data["Fare"].max() - data["Fare"].min()

  # The number of passengers with cabins
  cabin_pass = data["Cabin"].count()

  # The code for the port where the highest number of passengers embarked
  port_highest_embark = data["Embarked"].mode()

  # The most populous gender
  most_populous_gender = data["Sex"].mode()

  # The standard deviation for age and fare
  std_deviation_age_fare = data[["Age", "Fare"]].std()


  return {
      "The total number of passengers on the titanic: ": tot_pass,
      "The age of the youngest passenger: ": youngest_pass,
      "The most expensive ticket price: ": most_exp_ticket,
      "The range of ticket prices: ": ticket_price_ranges,
      "The number of passengers with cabins: ": cabin_pass,
      "The code for the port where the highest number of passengers embarked: ": port_highest_embark[0],
      "The most populous gender: ": most_populous_gender[0],
      "The standard deviation for age and fare: ": std_deviation_age_fare,
  }

statistics_4 = investigate_data(data)

print("\nExercise 4 - displaying other statistics\n")
for key, value in statistics_4.items():
  print(f"{key}:\n{value}\n")



Exercise 4 - displaying other statistics

The total number of passengers on the titanic: :
891

The age of the youngest passenger: :
0.42

The most expensive ticket price: :
512.3292

The range of ticket prices: :
512.3292

The number of passengers with cabins: :
204

The code for the port where the highest number of passengers embarked: :
S

The most populous gender: :
male

The standard deviation for age and fare: :
Age     14.526497
Fare    49.693429
dtype: float64



### Exercise 5 - aggregating statistics grouped by category
---

Refer again to the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)   
looking particularly at the section on Aggregating statistics grouped by category.

1.  What is the mean age for male versus female Titanic passengers?
2.  What is the mean ticket fare price for each of the sex and cabin class combinations?
3.  What is the mean ticket fare price for passengers who embarked at each port?
4.  Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?

**Test output**:  
1.  female 27.915709 male 30.726645
2.  
```
female  1         106.125798
            2          21.970121
            3          16.118810
male    1          67.226127
            2          19.741782
            3          12.661633
```
3.  
```
C    59.954144
Q    13.276030
S    27.079812
```
4.

```
Survived  Pclass

0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
```






In [35]:
def investigate_data(data):
  # 1. What is the mean age for male versus female Titanic passengers?
  mean_age_by_gender = data[["Sex", "Age"]].groupby("Sex").mean()

  # 2. What is the mean ticket fare price for each of the sex and cabin class combinations?
  mean_fare_by_sex_cabin = data.groupby(["Sex", "Pclass"])["Fare"].mean()

  # 3. What is the mean ticket fare price for passengers who embarked at each port?
  mean_fare_by_port = data.groupby(["Embarked"])["Fare"].mean()

  # 4. Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?
  survival_by_pclass = data.groupby(["Survived", "Pclass"])["Survived"].count()

  return {
    "1. What is the mean age for male versus female Titanic passengers?: ": mean_age_by_gender,
    "2. What is the mean ticket fare price for each of the sex and cabin class combinations?: ": mean_fare_by_sex_cabin,
    "3. What is the mean ticket fare price for passengers who embarked at each port?: ": mean_fare_by_port,
    "4. Which passenger class had the highest number of survivors?: ": survival_by_pclass,
  }

statistics_5 = investigate_data(data)

print("\nExercise 5 - aggregating statistics grouped by category\n")
for key, value in statistics_5.items():
  print(f"{key}:\n{value}\n")



Exercise 5 - aggregating statistics grouped by category

1. What is the mean age for male versus female Titanic passengers?: :
              Age
Sex              
female  27.915709
male    30.726645

2. What is the mean ticket fare price for each of the sex and cabin class combinations?: :
Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64

3. What is the mean ticket fare price for passengers who embarked at each port?: :
Embarked
C    59.954144
Q    13.276030
S    27.079812
Name: Fare, dtype: float64

4. Which passenger class had the highest number of survivors?: :
Survived  Pclass
0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
Name: Survived, dtype: int64



### Exercise 6 - an aggregation of different statistics
---

Use the function titanic.agg() as shown in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

1.  Display

```
     {
         "Age": ["min", "max", "median", "skew"],
         "Fare": ["min", "max", "median", "mean"]
     }
```
2.  Display:  
min, max and mean for Age  
min, max and standard deviation for Fare  
count for Cabin

**Test output**:   
1.
```
                  Age	      Fare  
max	   80.000000	512.329200  
mean	  NaN	      32.204208  
median	28.000000	14.454200  
min	   0.420000	 0.000000  
skew	  0.389108	 NaN
```

2.   
```
	        Age	    Fare	   Cabin
count  NaN	    NaN	    204.0
max	80.000000  512.329200 NaN
mean   29.699118  NaN        NaN
min	0.420000   0.000000   NaN
std	NaN        49.693429  NaN
```




In [36]:
# Exercise 6 - an aggregation of different statistics
# 1.
aggregation_1 = data.agg(
    {
      "Age": ["max", "median", "min", "skew"],
      "Fare": ["max", "mean", "median", "min"]
    }
)

# Define new index order
new_index_1 = ["max", "mean", "median", "min","skew"]

# Reorder rows/index
aggr_1_reindexed = aggregation_1.reindex(new_index_1)

print(f"\nExercise 6.1:\n {aggr_1_reindexed}\n")

# 2.
aggregation_2 = data.agg(
    {
      "Age": ["min", "max", "mean"],
      "Fare": ["min","max", "std"],
      "Cabin": ["count"]
    }
)

# Define new index order
new_index_2 = ["count","max","mean","min","std"]

# Reorder rows/index
aggr_2_reindexed = aggregation_2.reindex(new_index_2)

print(f"\nExercise 6.2:\n {aggr_2_reindexed}\n")


Exercise 6.1:
               Age        Fare
max     80.000000  512.329200
mean          NaN   32.204208
median  28.000000   14.454200
min      0.420000    0.000000
skew     0.389108         NaN


Exercise 6.2:
              Age        Fare  Cabin
count        NaN         NaN  204.0
max    80.000000  512.329200    NaN
mean   29.699118         NaN    NaN
min     0.420000    0.000000    NaN
std          NaN   49.693429    NaN



### Exercise 7 - count by category
---

Read the section Count number of records by category in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)

1. Display the number of passengers of each gender who had a ticket
2. Display the number of passengers who embarked at each port and had a ticket
3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic (*Hint:  try getting the PassengerIds with a count for survived or not.  Store this value in a new variable, which will contain a list/array.  The second item in this list will be the number who survived.  You can use this number and the count of PassengerIds to calculate the percentage*)

**Test output**:  
1.  female 314, male 577
2.  C 168, Q 77, S 644
3.  38.38383838383838



In [81]:
import pandas as pd

# Import CSV-data from GitHub
def get_csv_data():
  url = " https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

def investigate_data(data):
  # 1. Display the number of passengers of each gender who had a ticket:
  ticket_by_gender = data.groupby(["Sex"])["Ticket"].count()

  # 2. Display the number of passengers who embarked at each port and had a ticket:
  no_embarked = data.groupby(["Embarked"])["Ticket"].count()

  # 3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic:
  # Count Passengers
  no_of_pass = data["PassengerId"].count()

  # Count Survived
  survived = data["Survived"].value_counts()

  # Calculate Survival Percentage
  survival_percentage = (survived[1] / no_of_pass) * 100

  return {
    "1. Display the number of passengers of each gender who had a ticket: ": ticket_by_gender,
    "2. Display the number of passengers who embarked at each port and had a ticket: ": no_embarked,
    "3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic: ": survival_percentage,
  }

statistics_7 = investigate_data(data)


print("\nExercise 7 - count by category\n")
for key, value in statistics_7.items():
  print(f"{key}:\n{value}\n")


Exercise 7 - count by category

1. Display the number of passengers of each gender who had a ticket: :
Sex
female    314
male      577
Name: Ticket, dtype: int64

2. Display the number of passengers who embarked at each port and had a ticket: :
Embarked
C    168
Q     77
S    644
Name: Ticket, dtype: int64

3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic: :
38.38383838383838



### Exercise 8 - summary happiness statistics
---

Open the data set here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true

It contains data on people's perception of happiness levels in a number of countries across the world.

1.  Display the number of records in the set  
2.  Display the description of the numerical data  
3.  Display the highest GDP and life expectancy  
4.  Display the mean, max and min for Freedom,  mean, max, min and skew for Generosity and mean, max, min and std for GDP  

**Test output**:  
1.  156
2.  Table showing count, mean, std, min, 25%, 50%, 75%, max for 8 columns
3.  GDP 1.684, life expectancy 1.141  
4.  


```
	   Freedom to make life choices	Generosity	GDP per capita
max	 0.631000	                   0.566000	  1.684000
mean	0.392571	                   0.184846	  0.905147
min	 0.000000	                   0.000000  	0.000000
skew	NaN	                        0.745942	  NaN
std	 NaN	                        NaN	       0.398389
```




### Exercise 9 - migration data
---

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true  Open the sheet named *Country Migration*

1.  Describe the dataset  
2.  Show summary information
3.  Display the mean net per 10K migration in each of the years 2015 to 2019
4.  Display the mean, max and min migration for the year 2019 for each of the regions (*base_country_wb_region*)
5.  Display the median net migration for the years 2015 and 2019 for the base countries by income level
6.  Display the number of target countries in each income level
7.  Display the mean net migration for all five years, for each income level

**Test output**:  
1  count, mean, std, min, 25%, 50%, 75%, max for 9 columns

2  shows 16 columns with non-null count of 4148 in each column

3  
```
net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64
```

4  
```
	net_per_10K_2019
mean	max	min
base_country_wb_region
East Asia & Pacific	0.198827	21.57	-9.88
Europe & Central Asia	0.208974	87.71	-21.34
Latin America & Caribbean	-0.904602	21.15	-31.75
Middle East & North Africa	-0.107655	55.60	-50.33
North America	0.239246	23.20	-0.29
South Asia	-0.514577	13.72	-24.89
Sub-Saharan Africa	-0.279729	37.11	-21.54
```

5  
```
	net_per_10K_2015	net_per_10K_2019
base_country_wb_income
High Income	0.02	0.04
Low Income	0.42	-0.05
Lower Middle Income	-0.02	-0.07
Upper Middle Income	-0.03	-0.08
```

6  
```
base_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64
```

7  
```
net_per_10K_2015	net_per_10K_2016	net_per_10K_2017	net_per_10K_2018	net_per_10K_2019
base_country_wb_income
High Income	0.505482	0.391379	0.314178	0.379201	0.401470
Low Income	1.876432	0.798270	-0.684865	-0.677784	-0.681459
Lower Middle Income	0.591654	-0.029893	-0.519433	-0.527136	-0.476616
Upper Middle Income	-0.043419	-0.502916	-0.699240	-0.686626	-0.700101
```






### Exercise 10 - calculating range over a grouped series

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named *Skill Migration*

1.  Display the max for each skill group category of net migration for the year 2017
2.  Assign the max for each skill group category of net migration for the year 2017 to a variable called **max_skill_migration** and print `max_skill_migration`
3.  Create a second variable called **min_skill_migration** and assign to it the min for each skill group category of net migration for the year 2017, print `min_skill_migration`

4.  You now have two series `max_skill_migration` and `min_skill_migration` each of which is a numpy array.  You can perfom calculations on these two series in the same way as you would individual data items.

So, you can calculate the range for each skill category by subtracting the `min_skill_migration` from `max_skill_migration` to get a new series **skill_migration_range**

skill_migration_range = max_skill_migration - min_skill_migration

Try it out.

5.  Now calculate the range for the year 2019
6.  Now calculate the range for countries grouped by base country income level for the year 2015

Test output:  
1 and 2  
```
skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
```

3    
```
skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
```

4  
```
skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64
```

5  
```
skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64
```

6  
```
wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64
```



